### PR TITLE
Fix incorrect sequence length for rnn

### LINKED ISSUE #23202: [Bug] RLlib: Dict environment space with lstm wrapping and preprocessing disabled setting incorrect seq_lens
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### Issue Severity

High: It blocks me to complete my task.

### What happened + What you expected to happen

Dict environment space with lstm wrapping and preprocessing disabled setting incorrect seq_lens inside TorchPolicy.compute_actions_from_input_dict.

Run the reproduction script for more detailed understanding.

### Versions / Dependencies

Ray: 1.11.0
torch: 1.10.2
tensorflow: 2.8.0
Python: 3.9.7
OS: Linux Mint 20.3 with xanmod kernel 5.15.27

### Reproduction script

A modified example
```py
"""Example of using variable-length Repeated / struct observation spaces.

This example shows:
  - using a custom environment with Repeated / struct observations
  - using a custom model to view the batched list observations

For PyTorch / TF eager mode, use the `--framework=[torch|tf2|tfe]` flag.
"""

import argparse
import os

import ray
from ray import tune
from ray.rllib.models import ModelCatalog

from ray.rllib.models.torch.torch_modelv2 import TorchModelV2
from ray.rllib.utils.framework import try_import_torch
from ray.rllib.models.torch.fcnet import FullyConnectedNetwork as TorchFCNet

torch, nn = try_import_torch()


class CustomTorchRPGModel(TorchModelV2, nn.Module):
    """Example of interpreting repeated observations."""

    def __init__(self, obs_space, action_space, num_outputs, model_config,
                 name):
        super().__init__(obs_space, action_space, num_outputs, model_config,
                         name)
        nn.Module.__init__(self)
        print("[KINAL DEBUG]", obs_space, action_space, num_outputs, model_config, name)
        new_obs_space = (obs_space['status'])
        print("[KINAL DEBUG]", new_obs_space)
        self.model = TorchFCNet(new_obs_space, action_space, 64,
                                model_config, name)
        self.num_outputs = 64

    def forward(self, input_dict, state, seq_lens):
        input_dict["obs_flat"] = input_dict['obs']['status']
        return self.model.forward(input_dict, state, seq_lens)

    def value_function(self):
        return self.model.value_function()


import gym
from gym.spaces import Discrete, Box, Dict

from ray.rllib.utils.spaces.repeated import Repeated


class SimpleRPG(gym.Env):

    def __init__(self, config):
        self.cur_pos = 0
        self.action_space = Discrete(4)

        # Represents an item.
        self.item_space = Discrete(5)

        # Represents an effect on the player.
        self.effect_space = Box(9000, 9999, shape=(4, ))

        # Represents a player.
        self.player_space = Dict({
            "location": Box(-100, 100, shape=(2, )),
            "status": Box(-1, 1, shape=(10, )),
        })
        self.observation_space = self.player_space

    def reset(self):
        return self.observation_space.sample()

    def step(self, action):
        return self.observation_space.sample(), 1, True, {}


parser = argparse.ArgumentParser()
parser.add_argument(
    "--framework",
    choices=["tf", "tf2", "tfe", "torch"],
    default="torch",
    help="The DL framework specifier.",
)

from ray.rllib.agents.registry import get_trainer_class

if __name__ == "__main__":
    ray.init()
    args = parser.parse_args()
    ModelCatalog.register_custom_model("my_model", CustomTorchRPGModel)

    config = {
        "framework": args.framework,
        "env": SimpleRPG,
        "rollout_fragment_length": 1,
        "train_batch_size": 2,
        # Use GPUs iff `RLLIB_NUM_GPUS` env var set to > 0.
        "num_gpus": int(os.environ.get("RLLIB_NUM_GPUS", "0")),
        "num_workers": 0,
        "model": {
            "custom_model": "my_model",
            "use_lstm":True,
            "max_seq_len":10,
            "lstm_cell_size":20,
            "lstm_use_prev_action":True,
        },
        "_disable_preprocessor_api": True,

    }

    stop = {
        "timesteps_total": 1,
    }
    # trainer = get_trainer_class("IMPALA")(env=SimpleRPG, config=config)
    tune.run("PG", config=config, stop=stop, verbose=1)
```

### Anything else

The problem can be solved by changing the [line](https://github.com/ray-project/ray/blob/3625c4760ff8b30aad5461ad94838cbf6c3daa77/rllib/policy/torch_policy.py#L318) in file `ray/rllib/policy/torch_policy.py` to use `state_batches[0]` to generate sequence length instead of using `input_dict["obs"]` for the class method `TorchPolicy.compute_actions_from_input_dict`

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by kinalmehta**
## Why are these changes needed?
Update the torch policy to find the seq_lens using state_batches instead of input_dict. This helps handle the complex inputs to the model when the inbuilt preprocessing API is disabled.

## Related issue number
Closes #23202 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/policy/torch_policy.py [hunk 1]
```diff
@@ -317,9 +317,9 @@ def compute_actions_from_input_dict(
             # Calculate RNN sequence lengths.
             seq_lens = (
                 torch.tensor(
-                    [1] * len(input_dict["obs"]),
+                    [1] * len(state_batches[0]),
```
Comment (avnishn) [inline rllib/policy/torch_policy.py:320]: is this also a bug in the tf dynamic and eager policies? Could you inspect that and make a similar change for those if so? thanks!

**Review by avnishn — COMMENTED** [review]
I have a question about whether the bug applies to tf policies. Thanks for contributing @kinalmehta

File: rllib/policy/torch_policy.py (hunk 1 above)
Comment (kinalmehta) [reply rllib/policy/torch_policy.py:320]: I verified the code for eager_tf_policy.py and it is handling the seq_lens parameter correctly.
You can verify it here: https://github.com/ray-project/ray/blob/7d262f886dec5c68fe3270aea1492aa483048a68/rllib/policy/eager_tf_policy.py#L758

Thanks :D

**Review by gjoliver — APPROVED** [review]

**Comment by gjoliver** [comment]
thanks a lot for helping fix this!

**Pull request merged**